Implement extra cursors on macOS

JonnyPtn opened the pull request:
Addresses #1401

The move cursor doesn't look quite as I'd expect (It's basically just a white arrow) but I can't see any better looking alternatives.

mantognini commented:
SFMLBuildMaster: Build this please.

This looks good but haven't actually tested it. I'll try to do it this week. Thanks for the PR!

mantognini commented:
I guess the documentation should be updated as well, specifically these lines:

https://github.com/JonnyPtn/SFML/blob/76f3adbd978e8955d24dcc787339563edc307b90/include/SFML/Window/Cursor.hpp#L65-L67 

What about replacing *no* with *maybe (1)* and adding a note along those lines: *(1) those cursors are usually available on macOS \<insert minimal version\> and newer but without any guarantee.* (Or phrased completely differently...)

JonnyPtn commented:
Ah yes, Sorry I missed the docs, will update now. 

Not really sure what to put for the supported version, can't find any documentation about it, and looking at other projects doesn't yield any useful results (glfw has an open (issue)[https://github.com/glfw/glfw/issues/427], and as far as I can tell SDL just uses the closed hand cursor for these diagonal resizes (?)

JonnyPtn commented on src/SFML/Window/OSX/CursorImpl.mm:
```diff
@@ -80,6 +90,14 @@
+        case Cursor::SizeTopLeftBottomRight:
+            return loadFromSelector(@selector(_windowResizeNorthWestSouthEastCursor), m_cursor);
+        case Cursor::SizeAll:
+            return loadFromSelector(@selector(_moveCursor), m_cursor);
```
Will returning here cause issues as it doesn't hit `[m_cursor retain]`?

JonnyPtn commented:
I've added the Help and Wait cursors, and updated documentation as best I can.

You can use this test code to cycle through the new cursors with Space:
```
#include <SFML/Graphics.hpp>

int main(int, char const**)
{
    // Load cursors
    std::vector<sf::Cursor> cursors(5);
    
    cursors[0].loadFromSystem(sf::Cursor::SizeBottomLeftTopRight);
    cursors[1].loadFromSystem(sf::Cursor::SizeTopLeftBottomRight);
    cursors[2].loadFromSystem(sf::Cursor::SizeAll);
    cursors[3].loadFromSystem(sf::Cursor::Wait);
    cursors[4].loadFromSystem(sf::Cursor::Help);
    
    int currentCursor = -1;
    
    // Create the main window
    sf::RenderWindow window(sf::VideoMode(800, 600), "SFML window");

    // Start the game loop
    while (window.isOpen())
    {
        // Process events
        sf::Event event;
        while (window.pollEvent(event))
        {
            // Close window: exit
            if (event.type == sf::Event::Closed)
            {
                window.close();
            }

            // Escape pressed: exit
            if (event.type == sf::Event::KeyPressed && event.key.code == sf::Keyboard::Escape)
            {
                window.close();
            }
            if (event.type == sf::Event::KeyPressed && event.key.code == sf::Keyboard::Space)
            {
                currentCursor++;
                if (currentCursor >= cursors.size())
                    currentCursor = 0;
                window.setMouseCursor(cursors[currentCursor]);
            }
        }

        // Clear screen 
        window.clear();
    }

    return EXIT_SUCCESS;
}

```

mantognini commented:
Yes, most likely. Good catch!

On Mon, 16 Apr 2018, 18:49 Jonny Paton, <notifications@github.com> wrote:

> *@JonnyPtn* commented on this pull request.
> ------------------------------
>
> In src/SFML/Window/OSX/CursorImpl.mm
> <https://github.com/SFML/SFML/pull/1413#discussion_r181811767>:
>
> > @@ -80,6 +90,14 @@
>          case Cursor::SizeVertical:    m_cursor = [NSCursor resizeUpDownCursor];        break;
>          case Cursor::Cross:           m_cursor = [NSCursor crosshairCursor];           break;
>          case Cursor::NotAllowed:      m_cursor = [NSCursor operationNotAllowedCursor]; break;
> +
> +        // These cursor types are undocumented, may not be available on some platforms
> +        case Cursor::SizeBottomLeftTopRight:
> +            return loadFromSelector(@selector(_windowResizeNorthEastSouthWestCursor), m_cursor);
> +        case Cursor::SizeTopLeftBottomRight:
> +            return loadFromSelector(@selector(_windowResizeNorthWestSouthEastCursor), m_cursor);
> +        case Cursor::SizeAll:
> +            return loadFromSelector(@selector(_moveCursor), m_cursor);
>
> Will returning here cause issues as it doesn't hit [m_cursor retain]?
>
> —
> You are receiving this because your review was requested.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/SFML/SFML/pull/1413#pullrequestreview-112500934>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAqU60O5DZrc_5EYeNZ7Rfgyo6kNTE46ks5tpMupgaJpZM4TPVW2>
> .
>

mantognini commented:
OK, that's good. But we should make it clear that Apple can remove those in
the future.

I'll see to test this tomorrow or the day after. :)

On Mon, 16 Apr 2018, 19:22 Bruno Van de Velde, <notifications@github.com>
wrote:

> I tested the _windowResizeNorthWestSouthEastCursor on OS X Mountain Lion
> (10.8) and it works. So I don't think we really have to worry about systems
> where the cursor doesn't exist.
>
> —
> You are receiving this because your review was requested.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/SFML/SFML/pull/1413#issuecomment-381682005>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAqU6zkfDQHamPC9QnTN1Lvj-koLWy11ks5tpNM6gaJpZM4TPVW2>
> .
>

mantognini commented:
Superseded by #1425.
